packages/avocode-email-tagsinput/test/e2e/basic.spec.js

```diff
@@ -0,0 +1,116 @@
+import puppeteer from 'puppeteer'
+import { expect } from 'chai'
+import {
+  URL,
+  tagSubmitKeys,
+  tagsInputFieldSelector,
+  getAnchorAndFocus,
+  addEmailTag,
+  assertEmailTagCount,
+  assertInputText,
+} from './test-utils'
+
+describe('Basic Email Tags Input', () => {
+  let browser
+  let page
+
+  beforeAll(async () => {
+    browser = await puppeteer.launch(browserOptions)
+  })
+
+  beforeEach(async () => {
+    page = await browser.newPage()
+    await page.goto(URL)
+  })
+
+  afterEach(async () => {
+    await page.close()
+  })
+
+  afterAll(async () => {
+    await browser.close()
+  })
+
+  it('should render editor', async () => {
+    const editor = await page.waitForSelector(tagsInputFieldSelector)
+    expect(editor).to.exist
+  })
+
+  describe('user typing', () => {
+    it('should insert text', async () => {
+      await page.click(tagsInputFieldSelector)
+      await page.type(tagsInputFieldSelector, 'abcd')
+      await assertInputText(page, 'abcd')
+    })
+
+    it('should delete text with backspace', async () => {
+      await page.click(tagsInputFieldSelector)
+      await page.type(tagsInputFieldSelector, 'abcd')
+      await page.keyboard.press('Backspace')
+      await assertInputText(page, 'abc')
+    })
+
+    it('should delete text with delete', async () => {
+      await page.click(tagsInputFieldSelector)
+      await page.type(tagsInputFieldSelector, 'abcd')
+      await page.keyboard.press('ArrowLeft')
+      await page.keyboard.press('Delete')
+      await assertInputText(page, 'abc')
+    })
+
+    it('should navigate one char left', async () => {
+      await page.click(tagsInputFieldSelector)
+      await page.type(tagsInputFieldSelector, 'abcd')
+      await page.keyboard.press('ArrowLeft')
+      const [ anchorOffset ] = await getAnchorAndFocus(page)
+      expect(anchorOffset).to.equal(3)
+    })
+  })
+
+  describe('user typing different email edge cases', () => {
+    const unsupportedEmailEdgeCases = [
+      'testemail',
+      'testemail@ex',
+      'testemail@ex.11',
+      'double"123@quotes.co',
+      'withb[]g@bracket.com',
+      'withsm()ll@bracket.com',
+    ]
+    const supportedEmailEdgeCases = [
+      '1234@test.co',
+      'testemail@11.ex',
+      "single'123@quote.co",
+      'withcu{}ly@bracket.co',
+      'सिमप्लेफोल्देर$%#?&@nepali.co',
+    ]
+
+    for (const testEmail of unsupportedEmailEdgeCases) {
```

Comment:
Please do not create dynamic tests. I know you mean well and I should probably create a contribution guide or something but I am strongly opposed to these.

My reasons are that they don't really bring much value because the "spec" should be "specification" of test and clearly state what you are testing.

They are also harder to debug especially when they're implemented in the loops since adding a debugger now means stepping through other specs. Basically when test fails, I want to be able to jump to that specific failing spec and fix it.

Also the description does not really say why I am not allowed to add specific tag. There should be simply a spec like this:

`should not be allowed to add invalid email as a tag`

You do not need to test every combination of invalid emails, I don't see a point in it.